File: CS/WindowsApplication3/ColumnHeaderExtender.cs
using System;
using System.Collections.Generic;
using System.Windows.Forms;
using DevExpress.Skins;
using DevExpress.XtraGrid.Views.Grid;
using System.Drawing;
using DevExpress.XtraEditors.Drawing;
using DevExpress.XtraEditors.Controls;
using DevExpress.XtraGrid.Views.Grid.ViewInfo;
using DevExpress.XtraGrid.Drawing;
using DevExpress.Utils.Drawing;
using DevExpress.XtraEditors;
using DevExpress.XtraGrid.Columns;
using DevExpress.Utils;

namespace DXSample {
    public class ColumnHeaderExtender {
        GridView view;
        SkinEditorButtonPainter customButtonPainter;
        EditorButtonObjectInfoArgs args;
        Size buttonSize;

        public ColumnHeaderExtender(GridView view) {
            this.view = view;
            buttonSize = new Size(14, 14);
            
        }

        public void AddCustomButton() {
            CreateButtonPainter();
            CreateButtonInfoArgs();
            SubscribeToEvents();
        }

        private void CreateButtonInfoArgs() {
            EditorButton btn = new EditorButton(ButtonPredefines.Glyph);
            args = new EditorButtonObjectInfoArgs(btn, new DevExpress.Utils.AppearanceObject());
        }

        private void CreateButtonPainter() {
            customButtonPainter = new SkinEditorButtonPainter(DevExpress.LookAndFeel.UserLookAndFeel.Default.ActiveLookAndFeel);
        }

        private void SubscribeToEvents() {
            view.CustomDrawColumnHeader += OnCustomDrawColumnHeader;
            view.MouseDown += OnMouseDown;
            view.MouseUp += OnMouseUp;
            view.MouseMove += OnMouseMove;
        }

        void OnMouseUp(object sender, MouseEventArgs e) {
            GridHitInfo hitInfo = view.CalcHitInfo(e.Location);
            if(hitInfo.HitTest != GridHitTest.Column) return;
            GridColumn column = hitInfo.Column;
            if(IsButtonRect(e.Location, column)) {
                SetButtonState(column, ObjectState.Normal);
                XtraMessageBox.Show(string.Format("Custom Button in {0}", column.FieldName));
                // your code here
                DXMouseEventArgs.GetMouseArgs(e).Handled = true;
            }
        }

        void OnMouseMove(object sender, MouseEventArgs e) {
            GridHitInfo hitInfo = view.CalcHitInfo(e.Location);
            if(hitInfo.HitTest != GridHitTest.Column) return;
            GridColumn column = hitInfo.Column;
            if(IsButtonRect(e.Location, column))
                SetButtonState(column, ObjectState.Hot);
            else
                SetButtonState(column, ObjectState.Normal);
        }

        void OnMouseDown(object sender, MouseEventArgs e) {
            GridHitInfo hitInfo = view.CalcHitInfo(e.Location);
            if(hitInfo.HitTest != GridHitTest.Column) return;
            GridColumn column = hitInfo.Column;
            if(IsButtonRect(e.Location, column)) {
                SetButtonState(column, ObjectState.Pressed);
                DXMouseEventArgs.GetMouseArgs(e).Handled = true;
            }
        }

        private void SetButtonState(GridColumn column, ObjectState state) {
            column.Tag = state;
            view.InvalidateColumnHeader(column);
        }

        private bool IsButtonRect(Point point, GridColumn column) {
            GraphicsInfo info = new GraphicsInfo();
            info.AddGraphics(null);
            GridViewInfo viewInfo = view.GetViewInfo() as GridViewInfo;
            GridColumnInfoArgs columnArgs = viewInfo.ColumnsInfo[column];
            Rectangle buttonRect = CalcButtonRect(columnArgs, info.Graphics);
            info.ReleaseGraphics();
            return buttonRect.Contains(point);
        }

        private Rectangle CalcButtonRect(GridColumnInfoArgs columnArgs, Graphics gr) {
            Rectangle columnRect = columnArgs.Bounds;
            int innerElementsWidth = CalcInnerElementsMinWidth(columnArgs, gr);
            Rectangle buttonRect = new Rectangle(columnRect.Right - innerElementsWidth - buttonSize.Width - 2,
                columnRect.Y + columnRect.Height / 2 - buttonSize.Height / 2, buttonSize.Width, buttonSize.Height);
            return buttonRect;
        }

        private int CalcInnerElementsMinWidth(GridColumnInfoArgs columnArgs, Graphics gr) {
            bool canDrawMode = true;
            return columnArgs.InnerElements.CalcMinSize(gr, ref canDrawMode).Width;
        }

        void OnCustomDrawColumnHeader(object sender, ColumnHeaderCustomDrawEventArgs e) {
            if(e.Column == null) return;
            DefaultDrawColumnHeader(e);
            DrawCustomButton(e);
            e.Handled = true;
        }

        private void DrawCustomButton(ColumnHeaderCustomDrawEventArgs e) {
            SetUpButtonInfoArgs(e);
            customButtonPainter.DrawObject(args);
        }

        private void SetUpButtonInfoArgs(ColumnHeaderCustomDrawEventArgs e) {
            args.Cache = e.Cache;
            args.Bounds = CalcButtonRect(e.Info, e.Cache.Graphics);
            ObjectState state = ObjectState.Normal;
            if(e.Column.Tag is ObjectState)
                state = (ObjectState)e.Column.Tag;
            args.State = state;
        }

        private static void DefaultDrawColumnHeader(ColumnHeaderCustomDrawEventArgs e) {
            e.Painter.DrawObject(e.Info);
        }

        private void UnsubscribeFromEvents() {
            view.CustomDrawColumnHeader -= OnCustomDrawColumnHeader;
            view.MouseDown -= OnMouseDown;
            view.MouseUp -= OnMouseUp;
            view.MouseMove -= OnMouseMove;
        }

        public void RemoveCustomButton() {
            UnsubscribeFromEvents();

        }
    }
}

File: VB/WindowsApplication3/ColumnHeaderExtender.vb
Imports System
Imports System.Collections.Generic
Imports System.Windows.Forms
Imports DevExpress.Skins
Imports DevExpress.XtraGrid.Views.Grid
Imports System.Drawing
Imports DevExpress.XtraEditors.Drawing
Imports DevExpress.XtraEditors.Controls
Imports DevExpress.XtraGrid.Views.Grid.ViewInfo
Imports DevExpress.XtraGrid.Drawing
Imports DevExpress.Utils.Drawing
Imports DevExpress.XtraEditors
Imports DevExpress.XtraGrid.Columns
Imports DevExpress.Utils

Namespace DXSample
	Public Class ColumnHeaderExtender
		Private view As GridView
		Private customButtonPainter As SkinEditorButtonPainter
		Private args As EditorButtonObjectInfoArgs
		Private buttonSize As Size

		Public Sub New(ByVal view As GridView)
			Me.view = view
			buttonSize = New Size(14, 14)

		End Sub

		Public Sub AddCustomButton()
			CreateButtonPainter()
			CreateButtonInfoArgs()
			SubscribeToEvents()
		End Sub

		Private Sub CreateButtonInfoArgs()
			Dim btn As New EditorButton(ButtonPredefines.Glyph)
			args = New EditorButtonObjectInfoArgs(btn, New DevExpress.Utils.AppearanceObject())
		End Sub

		Private Sub CreateButtonPainter()
			customButtonPainter = New SkinEditorButtonPainter(DevExpress.LookAndFeel.UserLookAndFeel.Default.ActiveLookAndFeel)
		End Sub

		Private Sub SubscribeToEvents()
			AddHandler view.CustomDrawColumnHeader, AddressOf OnCustomDrawColumnHeader
			AddHandler view.MouseDown, AddressOf OnMouseDown
			AddHandler view.MouseUp, AddressOf OnMouseUp
			AddHandler view.MouseMove, AddressOf OnMouseMove
		End Sub

		Private Sub OnMouseUp(ByVal sender As Object, ByVal e As MouseEventArgs)
			Dim hitInfo As GridHitInfo = view.CalcHitInfo(e.Location)
			If hitInfo.HitTest <> GridHitTest.Column Then
				Return
			End If
			Dim column As GridColumn = hitInfo.Column
			If IsButtonRect(e.Location, column) Then
				SetButtonState(column, ObjectState.Normal)
				XtraMessageBox.Show(String.Format("Custom Button in {0}", column.FieldName))
				' your code here
				DXMouseEventArgs.GetMouseArgs(e).Handled = True
			End If
		End Sub

		Private Sub OnMouseMove(ByVal sender As Object, ByVal e As MouseEventArgs)
			Dim hitInfo As GridHitInfo = view.CalcHitInfo(e.Location)
			If hitInfo.HitTest <> GridHitTest.Column Then
				Return
			End If
			Dim column As GridColumn = hitInfo.Column
			If IsButtonRect(e.Location, column) Then
				SetButtonState(column, ObjectState.Hot)
			Else
				SetButtonState(column, ObjectState.Normal)
			End If
		End Sub

		Private Sub OnMouseDown(ByVal sender As Object, ByVal e As MouseEventArgs)
			Dim hitInfo As GridHitInfo = view.CalcHitInfo(e.Location)
			If hitInfo.HitTest <> GridHitTest.Column Then
				Return
			End If
			Dim column As GridColumn = hitInfo.Column
			If IsButtonRect(e.Location, column) Then
				SetButtonState(column, ObjectState.Pressed)
				DXMouseEventArgs.GetMouseArgs(e).Handled = True
			End If
		End Sub

		Private Sub SetButtonState(ByVal column As GridColumn, ByVal state As ObjectState)
			column.Tag = state
			view.InvalidateColumnHeader(column)
		End Sub

		Private Function IsButtonRect(ByVal point As Point, ByVal column As GridColumn) As Boolean
			Dim info As New GraphicsInfo()
			info.AddGraphics(Nothing)
			Dim viewInfo As GridViewInfo = TryCast(view.GetViewInfo(), GridViewInfo)
			Dim columnArgs As GridColumnInfoArgs = viewInfo.ColumnsInfo(column)
			Dim buttonRect As Rectangle = CalcButtonRect(columnArgs, info.Graphics)
			info.ReleaseGraphics()
			Return buttonRect.Contains(point)
		End Function

		Private Function CalcButtonRect(ByVal columnArgs As GridColumnInfoArgs, ByVal gr As Graphics) As Rectangle
			Dim columnRect As Rectangle = columnArgs.Bounds
			Dim innerElementsWidth As Integer = CalcInnerElementsMinWidth(columnArgs, gr)
			Dim buttonRect As New Rectangle(columnRect.Right - innerElementsWidth - buttonSize.Width - 2, columnRect.Y + columnRect.Height \ 2 - buttonSize.Height \ 2, buttonSize.Width, buttonSize.Height)
			Return buttonRect
		End Function

		Private Function CalcInnerElementsMinWidth(ByVal columnArgs As GridColumnInfoArgs, ByVal gr As Graphics) As Integer
			Dim canDrawMode As Boolean = True
			Return columnArgs.InnerElements.CalcMinSize(gr, canDrawMode).Width
		End Function

		Private Sub OnCustomDrawColumnHeader(ByVal sender As Object, ByVal e As ColumnHeaderCustomDrawEventArgs)
			If e.Column Is Nothing Then
				Return
			End If
			DefaultDrawColumnHeader(e)
			DrawCustomButton(e)
			e.Handled = True
		End Sub

		Private Sub DrawCustomButton(ByVal e As ColumnHeaderCustomDrawEventArgs)
			SetUpButtonInfoArgs(e)
			customButtonPainter.DrawObject(args)
		End Sub

		Private Sub SetUpButtonInfoArgs(ByVal e As ColumnHeaderCustomDrawEventArgs)
			args.Cache = e.Cache
			args.Bounds = CalcButtonRect(e.Info, e.Cache.Graphics)
			Dim state As ObjectState = ObjectState.Normal
			If TypeOf e.Column.Tag Is ObjectState Then
				state = DirectCast(e.Column.Tag, ObjectState)
			End If
			args.State = state
		End Sub

		Private Shared Sub DefaultDrawColumnHeader(ByVal e As ColumnHeaderCustomDrawEventArgs)
			e.Painter.DrawObject(e.Info)
		End Sub

		Private Sub UnsubscribeFromEvents()
			RemoveHandler view.CustomDrawColumnHeader, AddressOf OnCustomDrawColumnHeader
			RemoveHandler view.MouseDown, AddressOf OnMouseDown
			RemoveHandler view.MouseUp, AddressOf OnMouseUp
			RemoveHandler view.MouseMove, AddressOf OnMouseMove
		End Sub

		Public Sub RemoveCustomButton()
			UnsubscribeFromEvents()

		End Sub
	End Class
End Namespace
